okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -1460,4 +1563,246 @@ public void alternateService(int streamId, String origin, ByteString protocol, S
       // TODO(madongfly): Deal with alternateService propagation
     }
   }
+
+  /**
+   * SSLSocket wrapper that provides a fake SSLSession for handshake session.
+   */
+  static class SslSocketWrapper extends SSLSocket {
+
+    private final SSLSession sslSession;
+    private final SSLSocket sslSocket;
+
+    SslSocketWrapper(SSLSocket sslSocket, String peerHost) {
+      this.sslSocket = sslSocket;
+      this.sslSession = new FakeSslSession(peerHost);
+    }
+
+    @Override
+    public SSLSession getHandshakeSession() {
+      return this.sslSession;
+    }
+
+    @Override
+    public boolean isConnected() {
+      return sslSocket.isConnected();
+    }
+
+    @Override
+    public SSLParameters getSSLParameters() {
+      SSLParameters sslParameters = sslSocket.getSSLParameters();
+      sslParameters.setEndpointIdentificationAlgorithm("HTTPS");
```

Comment:
Why was this needed?

Reply:
The endpointIdentificationAlgorithm in sslParameters of the real sslSocket is null. In the case of Netty we are setting it here https://github.com/grpc/grpc-java/blob/f8f613984fe1c57171977e948315449e3d013ebc/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L607 for the handshake but not for Okhttp.

Reply:
But changing here isn't actually changing it for the connection. I think we're generally depending on the HostnameVerifier instead. `setEndpointIdentificationAlgorithm()` was [added in Android API level 24](https://developer.android.com/reference/javax/net/ssl/SSLParameters#setEndpointIdentificationAlgorithm(java.lang.String)). We're claiming to support API level 21 (although [we should change that to 23](https://github.com/grpc/grpc-java/issues/11474))

I'm really bothered by the lack of failure building this. AnimalSniffer missed some things at times, but it seems to be completely broken right now. Running with `--console=plain` I see animalsnifferMain is SKIPPED. Digging deeper, it seems we're missing `@signature` in the version so it hasn't been running.

Reply:
But the setting endpoint algorithm in the wrapper SslSocket was for causing the hostname check to occur when we do the per rpc check using the X509ExrendedTrustManager.  Do we not want that to happen?

Reply:
As long as we don't have a minimum API level of 24 we can't check authority using X509ExtendedTrustManager. Should we then remove these changes and use only HostnameVerifier, and disallow per rpc authority override if one is not provided?